
The part of the program contributing to the running time the most
is the reading in of the reads. Here the inner loop taking the longest
is going through each kmer in each read and comparing it to the
kmers in the genes found in the "kmer2gene2kmerpos" datastructure.
This part has the following running time: \\

O(avg_read_len*number_of_genes*avg_dna_len) \\

Here the avg_dna_read is contributing to the outer loop 
going through each kmer in the "kmer_list." This is beacuse
the average read length (avg_read_len) is corresponding to the amount
of kmers found. \\
The number_of_genes_found is as the names suggest the number of genes found for each kmer.
This would in the worst case be all of them for each kmer, but this is unlikely. 
Here finding the kmers from the read which are present in the read is constant, since a dictonary 
is used in the "kmer2gene2kmerpos" datastructure to store the kmers.
Since the avg_read_len and the avg_dna_len can be considered
constant the inner loops running time could instead be considered: \\

O(number_of_genes) \\

Sine we are doing this for each read the total running time is: \\

O(number_of_reads * number_of_genes) \\

Here it could be argued that the number of genes is going to a lot smaller 
than the number of reads resulting in a linear running time in "number_of_reads"


